examples/flutter_gallery/lib/demo/animation/home.dart

```diff
@@ -0,0 +1,533 @@
+      final Section section = sections[index];
+      children.add(new LayoutId(
+        id: 'title$index',
+        child: new SectionTitle(
+          section: section,
+          scale: _titleScale(index),
+          opacity: _titleOpacity(index),
+        ),
+      ));
+    }
+
+    for(int index = 0; index < sections.length; index++) {
+      children.add(new LayoutId(
+        id: 'indicator$index',
+        child: new SectionIndicator(
+          opacity: _indicatorOpacity(index),
+        ),
+      ));
+    }
+
+    return new CustomMultiChildLayout(
+      delegate: new _AllSectionsLayout(
+        tColumnToRow: tColumnToRow,
+        tCollapsed: tCollapsed,
+        cardCount: sections.length,
+        selectedIndex: selectedIndex,
+      ),
+      children: children,
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return new LayoutBuilder(builder: _build);
+  }
+}
+
+class AnimationDemoHome extends StatefulWidget {
+  AnimationDemoHome({ Key key }) : super(key: key);
+
+  static const String routeName = '/animation';
+
+  @override
+  AnimationDemoHomeState createState() => new AnimationDemoHomeState();
+}
+
+class AnimationDemoHomeState extends State<AnimationDemoHome> {
+  double _selectedIndex = 0.0;
+  ScrollController _scrollController = new ScrollController();
+  PageController _headingPageController = new PageController();
+  PageController _detailsPageController = new PageController();
+
+  @override
+  Widget build(BuildContext context) {
+    return new Scaffold(
+      backgroundColor: _kAppBackgroundColor,
+      body: new Builder(
+        // Insert an element so that _buildBody can find the PrimaryScrollController.
+        builder: (BuildContext context) => _buildBody(context),
+      ),
+    );
+  }
+
+  void _maybeScroll(double midScrollOffset, int pageIndex, double xOffset) {
+    const Duration duration = const Duration(milliseconds: 400);
+    const Curve curve = Curves.fastOutSlowIn;
+    if (_scrollController.offset < midScrollOffset) {
+      // Scroll the overall list to the point where only one section card shows.
+      // At the same time scroll the PageViews to the page at pageIndex.
+      _headingPageController.animateToPage(pageIndex, curve: curve, duration: duration);
+      _scrollController.animateTo(midScrollOffset, curve: curve, duration: duration);
+    } else {
+      // One one section card is showing: scroll one page forward or back.
+      final double centerX = _headingPageController.position.viewportDimension / 2.0;
+      final int newPageIndex = xOffset > centerX ? pageIndex + 1 : pageIndex - 1;
+      _headingPageController.animateToPage(newPageIndex, curve: curve, duration: duration);
+    }
+  }
+
+  bool _handlePageNotification(ScrollNotification notification, PageController leader, PageController follower) {
+    if (notification.depth == 0 && notification is ScrollUpdateNotification) {
+      setState(() {
```

Comment:
Calling setState on every ScrollUpdateNotification is expensive because that will trigger this (big) widget to rebuild every frame of the scroll.  It would be more efficient to make _selectedIndex into an AnimationController and use SlideTransition rather than FractionalTranslation on line 456.  Similarly, _AllSectionsView could become an AnimatedWidget and listen to the animation as well.

(You'd never actually tell the _selectedIndex animation to go forward(), you'd just set the value and rely on that trigger the widgets listening to the animation.)